### PR TITLE
Add MidMeasureMP in LightningQubit (cpp only)

### PR DESCRIPTION
**Pull request by tomlqc**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Native-execution of circuits with mid-measurements is enabled via the dynamic-one-shot transform. To support the transform, a device must implement methods to apply MidMeasureMP.

MidMeasureMP is a special kind of operation where a sample is drawn from the state vector probability distribution and a projector is applied to the state vector accordingly. It may also support options like reset and post-selection.

The most straightforward way to deal with the requirements is to enhance the state vector class StateVectorLQubit.

This PR is only about the C++ part.

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

**Related GitHub Pull Request:**
https://github.com/PennyLaneAI/pennylane-lightning/pull/621 - See discussions

**Related Shortcut Story:**
[sc-57704]

### CODE REVIEW

**Review by AmintorDusko — COMMENTED** [review]
Can we have some tests for unit tests for normalize?

**Comment by tomlqc** [comment]
> Can we have some tests for unit tests for normalize?

Currently it's tested as part of `collapse()`. But, yes, of course we can have separated tests. I just would have preferred getting the feedback on the previous review.

**Comment by AmintorDusko** [comment]
> > Can we have some tests for unit tests for normalize?
> 
> Currently it's tested as part of `collapse()`. But, yes, of course we can have separated tests. I just would have preferred getting the feedback on the previous review.
> 
> Actually I may add the separate tests when I commit the Kokkos version, would this be ok?

@tomlqc, we didn't see it before. It is healthy for our software stack to have unit tests, that's how we get the small problems before they turn into big problems.
I will approve for now, and you can proceed with the tests in the next PR.

**Review by AmintorDusko — APPROVED** [review]

**Comment by tomlqc** [comment]
Thanks @AmintorDusko I understand your point. I'll make sure to add the unittest in the next PR.

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ### New features since last release
 
+* `lightning.qubit` supports mid-circuit measurements.
+  [(#621)](https://github.com/PennyLaneAI/pennylane-lightning/pull/621)
+
```
Comment (vincentmr) [inline .github/CHANGELOG.md:7]: The PR context is excellent, but I would suggest updating the changes and the changelog as follows
```suggestion
* Add `collapse` and `normalize` methods to the `StateVectorLQubit` classes, enabling "branching" of the wavefunction. Add methods to create and seed an RNG in the `Measurements` modules.
  [(#645)](https://github.com/PennyLaneAI/pennylane-lightning/pull/645)

```

**Review by vincentmr — APPROVED** [review]
Looks good to me, thanks @tomlqc .

**Pull request merged**